istioctl/cmd/vm-bootstrap.go

```diff
@@ -0,0 +1,630 @@
+}
+
+func parseRemoteResponse(reader io.Reader) (*remoteResponse, error) {
+	buffer := make([]uint8, 1)
+	if _, err := reader.Read(buffer); err != nil {
+		return nil, err
+	}
+
+	typ := buffer[0]
+	if typ > 0 {
+		buf := bufio.NewReader(reader)
+		message, err := buf.ReadString('\n')
+		if err != nil {
+			return nil, err
+		}
+		return &remoteResponse{typ, message}, nil
+	}
+
+	return &remoteResponse{typ: typ, message: ""}, nil
+}
+
+func checkRemoteResponse(r io.Reader) error {
+	response, err := parseRemoteResponse(r)
+	if err != nil {
+		return err
+	}
+
+	if response.typ > 0 {
+		return errors.New(response.message)
+	}
+
+	return nil
+}
+
+func waitTimeout(wg *sync.WaitGroup, timeout time.Duration) bool {
+	c := make(chan struct{})
+	go func() {
+		defer close(c)
+		wg.Wait()
+	}()
+	select {
+	case <-c:
+		return false // completed normally.
+	case <-time.After(timeout):
+		return true // timed out.
+	}
+}
+
+func remoteCopyFile(data []byte, location string, client *ssh.Client) error {
+	session, err := client.NewSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+
+	filename := path.Base(location)
+	r := bytes.NewReader(data)
+
+	wg := sync.WaitGroup{}
+	wg.Add(2)
+	errCh := make(chan error, 2)
+
+	size := len(data)
+
+	go func() {
+		defer wg.Done()
+		w, err := session.StdinPipe()
+		if err != nil {
+			errCh <- err
+			return
+		}
+		defer w.Close()
+
+		stdout, err := session.StdoutPipe()
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = fmt.Fprintln(w, "C0644", size, filename)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		if err = checkRemoteResponse(stdout); err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = io.Copy(w, r)
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		_, err = fmt.Fprint(w, "\x00")
+		if err != nil {
+			errCh <- err
+			return
+		}
+
+		if err = checkRemoteResponse(stdout); err != nil {
+			errCh <- err
+			return
+		}
+	}()
+
+	go func() {
+		defer wg.Done()
+		err := session.Run(fmt.Sprintf("%s -qt %s", scpPath, location))
+		if err != nil {
+			errCh <- err
+			return
+		}
+	}()
+
+	if waitTimeout(&wg, 60*time.Second) {
+		return fmt.Errorf("timeout uploading file")
+	}
+
+	close(errCh)
+	for err := range errCh {
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func remoteRunCommand(command string, client *ssh.Client) error {
+	session, err := client.NewSession()
+	if err != nil {
+		return err
+	}
+	defer session.Close()
+	return session.Run(command)
+}
+
+func dumpCertificates(directory string, addressCertMapping map[string]workloadEntryAddressKeys) error {
+	err := os.MkdirAll(directory, os.ModePerm)
+	if err != nil && !os.IsExist(err) {
+		return err
+	}
+	for address, certs := range addressCertMapping {
+		err = ioutil.WriteFile(path.Join(directory, "cert-"+address+".pem"), certs.Cert, 0644)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(path.Join(directory, "key-"+address+".pem"), certs.Key, 0644)
```

Comment:
Another bad practice. 

If you can ssh - run an ssh command to the VM to generate the key pair there and get a CSR ( key doesn't get dumped to a local filesystem where it may be accidentally backed up or read), and maybe pass the CSR to istiod - with a kube exec or a new service - so signing also happens in a trusted env.

Reply:
That works in the case of where we want to SSH to a machine, and I'm happy to chang ethat there. However, there are people who want to run this as part of CI job and want to just dump the certificates without actually transferring them. So the dumping codepath will want to exist for that.

Reply:
to generate a CSR, you need some verifiable identity in the CSR with which Istio can validate the VM workload. Where are you going to get that identity? you need to go to k8s to get the JWT or something else. and how will the VM go to k8s to get its JWTs if it doesn't have access to k8s in the first place?

Reply:
The identity of the admin - plus a policy - is how you validate. 

Istio signs CSR (now) based on 'identity' - you give a JWT token for my-serviceaccount.ns - you get back a cert with the same identity.

Most identity providers support the concept of 'impersonation' or delegation - you have a config indicating that admin.ns can impersonate any cert for *.ns ( and nothing else). 

You deploy the admin.ns certificate on ansible or CI/CD - and it'll be able to provision any VM,
but only in that namespace and with an audit. 

It is pretty common practice.